Let rpc._all_gather use default RPC timeout

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #44418 [WIP] Use streams from pool on RPC callees
* **#44983 Let rpc._all_gather use default RPC timeout**

`_all_gather` was converted from `_wait_all_workers` and inherited its
5 seconds fixed timeout. As `_all_gather` meant to support a broader
set of use cases, the timeout configuration should be more flexible.
This PR makes `rpc._all_gather` use the global default RPC timeout.

Differential Revision: [D23794383](https://our.internmc.facebook.com/intern/diff/D23794383)

mrshenli committed:
Let rpc._all_gather use default RPC timeout

`_all_gather` was converted from `_wait_all_workers` and inherited its
5 seconds fixed timeout. As `_all_gather` meant to support a broader
set of use cases, the timeout configuration should be more flexible.
This PR makes `rpc._all_gather` use the global default RPC timeout.

[ghstack-poisoned]

mrshenli committed:
Update on "Let rpc._all_gather use default RPC timeout"


`_all_gather` was converted from `_wait_all_workers` and inherited its
5 seconds fixed timeout. As `_all_gather` meant to support a broader
set of use cases, the timeout configuration should be more flexible.
This PR makes `rpc._all_gather` use the global default RPC timeout.

[ghstack-poisoned]

mrshenli committed:
Update on "Let rpc._all_gather use default RPC timeout"


`_all_gather` was converted from `_wait_all_workers` and inherited its
5 seconds fixed timeout. As `_all_gather` meant to support a broader
set of use cases, the timeout configuration should be more flexible.
This PR makes `rpc._all_gather` use the global default RPC timeout.

Differential Revision: [D23794383](https://our.internmc.facebook.com/intern/diff/D23794383)

[ghstack-poisoned]

mrshenli committed:
Update on "Let rpc._all_gather use default RPC timeout"


`_all_gather` was converted from `_wait_all_workers` and inherited its
5 seconds fixed timeout. As `_all_gather` meant to support a broader
set of use cases, the timeout configuration should be more flexible.
This PR makes `rpc._all_gather` use the global default RPC timeout.

Differential Revision: [D23794383](https://our.internmc.facebook.com/intern/diff/D23794383)

[ghstack-poisoned]

mrshenli committed:
Update on "Let rpc._all_gather use default RPC timeout"


`_all_gather` was converted from `_wait_all_workers` and inherited its
5 seconds fixed timeout. As `_all_gather` meant to support a broader
set of use cases, the timeout configuration should be more flexible.
This PR makes `rpc._all_gather` use the global default RPC timeout.

Differential Revision: [D23794383](https://our.internmc.facebook.com/intern/diff/D23794383)

[ghstack-poisoned]

mrshenli committed:
Update on "Let rpc._all_gather use default RPC timeout"


`_all_gather` was converted from `_wait_all_workers` and inherited its
5 seconds fixed timeout. As `_all_gather` meant to support a broader
set of use cases, the timeout configuration should be more flexible.
This PR makes `rpc._all_gather` use the global default RPC timeout.

Differential Revision: [D23794383](https://our.internmc.facebook.com/intern/diff/D23794383)

[ghstack-poisoned]

pritamdamania87 reviewed (approved):
Code changes look good, although the CI/lint failures do seem real.

rohan-varma commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -121,7 +121,14 @@ PyObject* rpc_init(PyObject* /* unused */) {
+              "_get_timeout",
+              // intentionally not releasing GIL to avoid context switch
+              [](const RpcAgent& self) {
+                return self.getRpcTimeout().count() / float(kToMilliseconds);
```
nit: there is also `kSecToMsConversion` which is 1000, does this serve the same purpose?

mrshenli replied:
consolidated all conversion const into the one in `rpc_agent.h`

rohan-varma commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -121,7 +121,14 @@ PyObject* rpc_init(PyObject* /* unused */) {
-              py::call_guard<py::gil_scoped_release>());
+              py::call_guard<py::gil_scoped_release>())
+          .def(
+              "_get_timeout",
```
Could we use the existing `get_rpc_timeout` which is on the RPC module? https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/rpc/init.cpp#L690

mrshenli replied:
Good  catch! I missed that one.

rohan-varma commented on torch/distributed/rpc/api.py:
```diff
@@ -163,8 +164,8 @@ def _all_gather(obj):
-    # Set a long enough timeout for all shutdown messages to be processed.
-    timeout = 5  # second
+    if timeout == UNSET_RPC_TIMEOUT:
+        timeout = _get_current_rpc_agent()._get_timeout()
```
If we can use it, `set_timeout = rpc.get_rpc_timeout()` might read simpler here.

rohan-varma commented on torch/distributed/rpc/api.py:
```diff
@@ -196,15 +197,19 @@ def _all_gather(obj):
-                        worker_name=follower_name, timeout=timeout
-                    )
-                )
+                failed_followers.append(follower_name)
```
nit, this was before this PR: the `as ex` above can be taken out since it is not used in the exception handling.

mrshenli committed:
Update on "Let rpc._all_gather use default RPC timeout"


`_all_gather` was converted from `_wait_all_workers` and inherited its
5 seconds fixed timeout. As `_all_gather` meant to support a broader
set of use cases, the timeout configuration should be more flexible.
This PR makes `rpc._all_gather` use the global default RPC timeout.

Differential Revision: [D23794383](https://our.internmc.facebook.com/intern/diff/D23794383)

[ghstack-poisoned]

mrshenli committed:
Update on "Let rpc._all_gather use default RPC timeout"


`_all_gather` was converted from `_wait_all_workers` and inherited its
5 seconds fixed timeout. As `_all_gather` meant to support a broader
set of use cases, the timeout configuration should be more flexible.
This PR makes `rpc._all_gather` use the global default RPC timeout.

Differential Revision: [D23794383](https://our.internmc.facebook.com/intern/diff/D23794383)

[ghstack-poisoned]

mrshenli committed:
Update on "Let rpc._all_gather use default RPC timeout"


`_all_gather` was converted from `_wait_all_workers` and inherited its
5 seconds fixed timeout. As `_all_gather` meant to support a broader
set of use cases, the timeout configuration should be more flexible.
This PR makes `rpc._all_gather` use the global default RPC timeout.

Differential Revision: [D23794383](https://our.internmc.facebook.com/intern/diff/D23794383)

[ghstack-poisoned]

mrshenli committed:
Update on "Let rpc._all_gather use default RPC timeout"


`_all_gather` was converted from `_wait_all_workers` and inherited its
5 seconds fixed timeout. As `_all_gather` meant to support a broader
set of use cases, the timeout configuration should be more flexible.
This PR makes `rpc._all_gather` use the global default RPC timeout.

Differential Revision: [D23794383](https://our.internmc.facebook.com/intern/diff/D23794383)

[ghstack-poisoned]
